File: swift/DesignPatterns/Behavioral/Strategy.swift
//
//  Strategy.swift
//  DesignPatterns
//
//  Created by Soe Min Thein on 19/03/2026.
//

import Foundation

// MARK: - Strategy
// Defines the interface for interchangeable algorithms.

protocol RouteStrategy {
    func calculateRoute(startPoint: String, endPoint: String)
}

// MARK: - Concrete Strategies
// Each strategy encapsulates a specific route calculation algorithm.

class CarRouteStrategy: RouteStrategy {
    func calculateRoute(startPoint: String, endPoint: String) {
        print("Calculating car route from \(startPoint) to \(endPoint)")
    }
}

class BusRouteStrategy: RouteStrategy {
    func calculateRoute(startPoint: String, endPoint: String) {
        print("Calculating bus route from \(startPoint) to \(endPoint)")
    }
}

class WalkRouteStrategy: RouteStrategy {
    func calculateRoute(startPoint: String, endPoint: String) {
        print("Calculating walk route from \(startPoint) to \(endPoint)")
    }
}

// MARK: - Context
// Maintains a reference to a strategy and delegates the work to it.

class NavigationContext {
    private var routeStrategy: RouteStrategy
    
    init(routeStrategy: RouteStrategy) {
        self.routeStrategy = routeStrategy
    }
    
    func setRouteStrategy(_ strategy: RouteStrategy) {
        routeStrategy = strategy
    }
    
    func calculateRoute(startPoint: String, endPoint: String) {
        routeStrategy.calculateRoute(startPoint: startPoint, endPoint: endPoint)
    }
}

// MARK: - Usage
// Swap strategies at runtime without changing the context.

func strategyExample() {
    let busRouteStrategy = BusRouteStrategy()
    let navigationContext = NavigationContext(routeStrategy: busRouteStrategy)
    navigationContext.calculateRoute(startPoint: "Start Point", endPoint: "End Point")
    
    let carRouteStrategy = CarRouteStrategy()
    navigationContext.setRouteStrategy(carRouteStrategy)
    navigationContext.calculateRoute(startPoint: "Start Point", endPoint: "End Point")
    
    let walkRouteStrategy = WalkRouteStrategy()
    navigationContext.setRouteStrategy(walkRouteStrategy)
    navigationContext.calculateRoute(startPoint: "Start Point", endPoint: "End Point")
}


// MARK: - Example 2: Sorting Strategy

// MARK: - Strategy

protocol SortStrategy {
    func sort(_ list: [Int])
}

// MARK: - Concrete Strategies

class BubbleSort: SortStrategy {
    func sort(_ list: [Int]) {
        print("BubbleSort applied.")
    }
}

class QuickSort: SortStrategy {
    func sort(_ list: [Int]) {
        print("QuickSort applied.")
    }
}

class MergeSort: SortStrategy {
    func sort(_ list: [Int]) {
        print("MergeSort applied.")
    }
}

// MARK: - Context

class Products {
    private let list: [Int]
    private var sortStrategy: SortStrategy

    init(list: [Int], sortStrategy: SortStrategy) {
        self.list = list
        self.sortStrategy = sortStrategy
    }

    func setSortStrategy(_ sortStrategy: SortStrategy) {
        self.sortStrategy = sortStrategy
    }

    func sort() {
        sortStrategy.sort(list)
    }

    func display() {
        print("Sorted array: ")
        for num in list {
            print(num)
        }
    }
}

// MARK: - Usage

func sortStrategyExample() {
    let list = [64, 34, 25, 12, 22, 11, 90]

    let products = Products(list: list, sortStrategy: BubbleSort())

    products.sort()
    products.display()

    products.setSortStrategy(QuickSort())
    products.sort()
    products.display()

    products.setSortStrategy(MergeSort())
    products.sort()
    products.display()
}

File: dart/design_patterns/lib/behavioral/strategy.dart
// Intent: Define a family of algorithms, encapsulate each one,
// and make them interchangeable at runtime.

// Strategy
// Defines the interface for interchangeable algorithms.

abstract interface class RouteStrategy {
  void calculateRoute(String startPoint, String endPoint);
}

// Concrete Strategies
// Each strategy encapsulates a specific route calculation algorithm.

class CarRouteStrategy implements RouteStrategy {
  @override
  void calculateRoute(String startPoint, String endPoint) {
    print('Calculating car route from $startPoint to $endPoint');
  }
}

class BusRouteStrategy implements RouteStrategy {
  @override
  void calculateRoute(String startPoint, String endPoint) {
    print('Calculating bus route from $startPoint to $endPoint');
  }
}

class WalkRouteStrategy implements RouteStrategy {
  @override
  void calculateRoute(String startPoint, String endPoint) {
    print('Calculating walk route from $startPoint to $endPoint');
  }
}

// Context
// Maintains a reference to a strategy and delegates the work to it.

class NavigationContext {
  RouteStrategy _routeStrategy;

  NavigationContext(RouteStrategy routeStrategy)
      : _routeStrategy = routeStrategy;

  void setRouteStrategy(RouteStrategy strategy) {
    _routeStrategy = strategy;
  }

  void calculateRoute(String startPoint, String endPoint) {
    _routeStrategy.calculateRoute(startPoint, endPoint);
  }
}

// Usage
// Swap strategies at runtime without changing the context.

void strategyExample() {
  final navigationContext = NavigationContext(BusRouteStrategy());
  navigationContext.calculateRoute('Start Point', 'End Point');

  navigationContext.setRouteStrategy(CarRouteStrategy());
  navigationContext.calculateRoute('Start Point', 'End Point');

  navigationContext.setRouteStrategy(WalkRouteStrategy());
  navigationContext.calculateRoute('Start Point', 'End Point');
}

// Example 2: Sorting Strategy

// Strategy

abstract interface class SortStrategy {
  void sort(List<int> list);
}

// Concrete Strategies

class BubbleSort implements SortStrategy {
  @override
  void sort(List<int> list) => print('BubbleSort applied.');
}

class QuickSort implements SortStrategy {
  @override
  void sort(List<int> list) => print('QuickSort applied.');
}

class MergeSort implements SortStrategy {
  @override
  void sort(List<int> list) => print('MergeSort applied.');
}

// Context

class Products {
  final List<int> _list;
  SortStrategy _sortStrategy;

  Products({required List<int> list, required SortStrategy sortStrategy})
      : _list = list,
        _sortStrategy = sortStrategy;

  void setSortStrategy(SortStrategy sortStrategy) {
    _sortStrategy = sortStrategy;
  }

  void sort() => _sortStrategy.sort(_list);

  void display() {
    print('Sorted array: ');
    for (final num in _list) {
      print(num);
    }
  }
}

// Usage

void sortStrategyExample() {
  final products = Products(
    list: [64, 34, 25, 12, 22, 11, 90],
    sortStrategy: BubbleSort(),
  );

  products.sort();
  products.display();

  products.setSortStrategy(QuickSort());
  products.sort();
  products.display();

  products.setSortStrategy(MergeSort());
  products.sort();
  products.display();
}

void main() {
  strategyExample();
  sortStrategyExample();
}
